### DISCUSSION TITLE
SPI speed set up using RadioLib

### DISCUSSION THREAD

**Discussion by Eugene-long** [jgromes/RadioLib#1168] [Q&A]
I am using an SX1268-based radio module by EBYTE (E22 400M22S) on an Arduino Nano. 5.1k resistors are used to match the signal levels. Therefore, the SPI speed should be reduced to 0.5 MHz. Is it possible to do this using RadioLib or should I use <SPI.h>?

**Comment by HeadBoffin** [comment]
This may help: https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup

**Reply by Eugene-long** [reply]
https://github.com/jgromes/RadioLib/wiki/Basics#non-standard-spi-setup :

Thank you for assistance. I do not need  to change the SPI bus pins. I  need  to change the SPI speed only.  Is it possible? 

*************************************************
Non-standard SPI setup
Some platforms (such as ESP32) allow user to change the SPI bus pins. By default, RadioLib is using the SPI instance provided by Arduino, and will automatically initialize the bus - usually by calling SPI.begin(). However, user may want to use a different SPI bus than the default, or do the initialization outside the library. For this cases, an extended Module class constructor is provided, in the format Module(cs, irq, rst, gpio, &spi, spiSettings), where spi is a reference to an instance of SPIClass, and spiSettings is an instance of class SPISettings. In such case, RadioLib will not perform any internal initialization of the bus, and therefore this is left up to the user, as shown in an example for SX1278 and ESP32 below.

SPIClass spi(VSPI);
SPISettings spiSettings(2000000, MSBFIRST, SPI_MODE0);
SX1278 radio = new Module(cs, irq, rst, gpio, spi, spiSettings);

void setup() {
  spi.begin();
  (...)
  radio.begin();
}
Please note that SPI bus MUST be initialized prior to calling the begin() method!
**********************************************************************

**Reply by jgromes** [reply]
@Eugene-long the same logic applies if you want to change some SPI settings, though I admit the wiki page is not entirely clear on that. I amended it to make it clearer, you will have to do something similar to what is shown below. Though I ma not sure Arduino Nano supports 500 kHz SPI speed to begin with.

```c++
SPISettings spiSettings(500000, MSBFIRST, SPI_MODE0);
SX1278 radio = new Module(cs, irq, rst, gpio, SPI, spiSettings);

void setup() {
  SPI.begin();
  (...)
  radio.begin();
}
````

**Reply by Eugene-long** [reply]
Hi Jan,
Thank you.

**Reply by Eugene-long** [reply]
Jan, Thank you once more.  Arduino Nano supports 500 kHz SPI speed. My LoRa system has started working.

**Comment by jgromes** [comment]
@Eugene-long 

> 5.1k resistors are used to match the signal levels. Therefore, the SPI speed should be reduced to 0.5 MHz.

I'm guessing you're using some resistor dividers to decrease the MOSI, CLK and NSS voltage levels from 5V to 3.3V. As @HeadBoffin pointed out, there's a wiki page that shows how to configure the SPI settings. However, just for my curiosity, where does the 0.5 MHz SPI speed come from though? The default 2 MHz speed should be slow enough for most cases.

Also, I would like to point out that the Arduino Nano is using ATmega328 microcontroller, which is rather limited in program and data memory, so I would not recommend it for new designs.
